pkg/chaosdaemon/tasks/task_config_manager.go

```diff
@@ -18,147 +18,163 @@ package tasks
+func (m TaskConfigManager) UpdateTaskConfig(id TaskID, task TaskConfig) (TaskConfig, error) {
 	if m.TaskConfigMap == nil {
 		return TaskConfig{}, ErrTaskConfigMapNotInit
 	}
 	taskOld, ok := m.TaskConfigMap[id]
 	if !ok {
-		return TaskConfig{}, errors.Wrapf(ErrUIDNotFound, "uid: %s, task: %v", id, task)
+		return TaskConfig{}, ErrNotFoundTaskID.WrapInput(id).WrapInput(task).Err()
 	}
-	if taskOld.main != task.main {
-		return TaskConfig{}, errors.Wrapf(ErrUpdateTaskConfigWithPIDChanges, "uid: %s, task: %v", id, task)
+	if taskOld.Main != task.Main {
+		return TaskConfig{}, ErrDiffID.Wrapf("expect: %v, input: %v", taskOld.Main, task.Main).Err()
 	}
 	m.TaskConfigMap[id] = task
 	return taskOld, nil
 }
 
 // DeleteTaskConfig Delete task inside the TaskConfigManager
-func (m TaskConfigManager) DeleteTaskConfig(id UID) error {
+func (m TaskConfigManager) DeleteTaskConfig(id TaskID) error {
 	if m.TaskConfigMap == nil {
 		return ErrTaskConfigMapNotInit
 	}
 	_, ok := m.TaskConfigMap[id]
 	if !ok {
-		return errors.Wrapf(ErrTaskConfigNotFound, "UID : %s", id)
+		return ErrNotFoundTypeTaskConfig.WrapInput(id).Err()
 	}
 	delete(m.TaskConfigMap, id)
 	return nil
 }
 
-func (m TaskConfigManager) GetConfigWithUID(id UID) (interface{}, error) {
+func (m TaskConfigManager) GetConfigWithUID(id TaskID) (TaskConfig, error) {
 	t, ok := m.TaskConfigMap[id]
 	if !ok {
-		return TaskConfig{}, ErrUIDNotFound
+		return TaskConfig{}, ErrNotFoundTaskID.WrapInput(id).Err()
 	}
-	return t.data, nil
+	return t, nil
 }
 
-func (m TaskConfigManager) GetUIDsWithPID(id PID) []UID {
-	uIds := make([]UID, 0)
+func (m TaskConfigManager) GetUIDsWithPID(id IsID) []TaskID {
+	uIds := make([]TaskID, 0)
 	for uid, task := range m.TaskConfigMap {
-		if task.main == id {
+		if task.Main == id {
 			uIds = append(uIds, uid)
 		}
 	}
 	return uIds
 }
 
-// SumTaskConfig will sum the TaskConfig with a same TaskConfig.main.
-// If developers want to use it with type T, they must implement Addable for *T.
-// IMPORTANT: Just here , we do not assume A.Add(B) == B.Add(A).
-// What SumTaskConfig do : A := new(TaskConfig), A.Add(B).Add(C).Add(D)... , A marked as uid.
-func (m TaskConfigManager) SumTaskConfig(uid UID) (TaskConfig, error) {
+func (m TaskConfigManager) CheckTask(uid TaskID, pid IsID) error {
+	t, ok := m.TaskConfigMap[uid]
+	if !ok {
+		return ErrNotFoundTaskID.WrapInput(uid).Err()
+	}
+	if t.Main != pid {
+		return ErrDiffID.Wrapf("expect: %v, input: %v", t.Main, pid).Err()
+	}
+	return nil
+}
+
+// MergeTaskConfig will sum the TaskConfig with a same TaskConfig.Main.
+// If developers want to use it with type T, they must implement Mergeable for *T.
+// IMPORTANT: Just here , we do not assume A.Merge(B) == B.Merge(A).
+// What MergeTaskConfig do : A := new(TaskConfig), A.Merge(B).Merge(C).Merge(D)... , A marked as uid.
+func (m TaskConfigManager) MergeTaskConfig(uid TaskID) (TaskConfig, error) {
```

Comment:
`MergeTaskConfig` is based on the `pid` right? It collects every config on a same `pid` and merge them together, but why it receives a `uid` and get the `pid` from it? Can we pass the `pid` directly to this function?

Reply:
It is because we can get a stable First Task base on this, meaning given two tasks uid A.uid & B.uid , the Merge order of MergeTaskConfig(A.uid) A.Merge(B) is differnt from the order of MergeTaskConfig(B.uid) : B.Merge(A).
If just give a pid we cannot divide these situation.

Reply:
So the `Merge` operation is not required to be commutative?

Reply:
Yes.

Reply:
No! As the golang map is not ordered, you cannot assume the order of the following merge (even though the first one is stable). For example, you may have `B.Merge(A).Merge(C)` or `B.Merge(C).Merge(A)`. The `Merge` has to be commutative to make them the same.

If we have the commutativity, can we use the `pid` instead of `uid` here?

Reply:
```
No! As the golang map is not ordered, you cannot assume the order of the following merge (even though the first one is stable). For example, you may have B.Merge(A).Merge(C) or B.Merge(C).Merge(A). The Merge has to be commutative to make them the same.
```

But I can make sure the first one , that was enough.

Reply:
Using `pid` means rejecting developer with incommutative config , I think it is not a good idea.

Reply:
Just for time-chaos in `clockIDsMask`. Developers can choose different policy on merging `clockIDsMask` base on this, but not just `clockIDsMask |= clockIDsMask` for every config.